src/bin/cobalt/serve.rs

```diff
@@ -78,6 +77,69 @@ pub fn serve_command(matches: &clap::ArgMatches) -> Result<()> {
     Ok(())
 }
 
+fn static_file_handler(dest: &path::Path, req: Request) -> Result<()> {
+    // grab the requested path
+    let mut req_path = req.url().to_string();
+
+    // strip off any querystrings so path.is_file() matches and doesn't stick index.html on the end
+    // of the path (querystrings often used for cachebusting)
+    if let Some(position) = req_path.rfind('?') {
+        req_path.truncate(position);
+    }
+
+    // find the path of the file in the local system
+    // (this gets rid of the '/' in `p`, so the `join()` will not replace the path)
+    let path = dest.to_path_buf().join(&req_path[1..]);
+
+    let serve_path = if path.is_file() {
+        // try to point the serve path to `path` if it corresponds to a file
+        path
+    } else {
+        // try to point the serve path into a "index.html" file in the requested
+        // path
+        path.join("index.html")
+    };
+
+    // if the request points to a file and it exists, read and serve it
+    if serve_path.exists() {
+        let mut file = fs::File::open(serve_path)?;
+        // buffer to store the file
+        let mut buffer: Vec<u8> = vec![];
+        file.read_to_end(&mut buffer)?;
+        req.respond(Response::from_data(buffer))?;
+    } else {
+        // write a simple body for the 404 page
+        req.respond(
+            Response::from_string("<h1> <center> 404: Page not found </center> </h1>")
+                .with_status_code(404),
+        )?;
+    }
+    Ok(())
+}
+
+fn serve(dest: &path::Path, ip: &str) -> Result<()> {
+    info!("Serving {:?} through static file server", dest);
+    info!("Server Listening on http://{}", &ip);
+    info!("Ctrl-c to stop the server");
+
+    // attempts to create a server
+    let server = match Server::http(ip) {
+        Ok(server) => server,
+        Err(e) => {
+            error!("{}", e);
+            process::exit(1);
+        }
+    };
```

Comment:
There a reason you don't propagate this up?

Reply:
I don't see any, especially as we process::exit. I think it was like this in cobalt before? I can change it to remove `error!()` :)

Reply:
The existing process exit is pretty much in response to ctrl-c since we don't need to print extra context and due to how some of the error reporting works..  In some of this code, errors might crop up for other reasons and those are ones we should propagate up to provide additional context.

Reply:
So should I left it there?

Reply:
I'd recommend getting rid of the exist's and see where that takes us

Reply:
However, if failing to create the HTTP server fails, we should exit out of the program.  We should do this with `?` so the error unwinds the stack and all of our context gets added as we do so.

Reply:
It's giving me hard time as the error is a Box<dyn std::error::Error + Send + Sync>, I can't just do `let server = Server::http(ip)?;` I need to do a `map_err` of some kind